### PR TITLE
Sparse accessor

### PR DESCRIPTION
**Pull request by mkrus**
data copying logic is flawed right now...

### CODE REVIEW

**Comment by mkrus** [comment]
fixed copying logic (thanks Juan!).  Need to add unit tests

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 1]
```diff
@@ -106,6 +109,74 @@ Qt3DRender::QAttribute::VertexBaseType accessorTypeFromJSON(int componentType)
...
+bool copySparseValues(int count,
+                      const QByteArray &indicesBuffer, int indicesOffset,
+                      const QByteArray &valuesBuffer, int valuesOffset,
+                      QByteArray &targetBuffer, int attributeOffset, int stride,
+                      Qt3DRender::QAttribute::VertexBaseType componentType, int numComponents)
+{
+    switch (componentType) {
+    case Qt3DRender::QAttribute::UnsignedByte:
+    case Qt3DRender::QAttribute::Byte:
+        return copySparseValues<IndexType>(count,
```
Comment (jjcasmar) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: do we really need all this switch again? you already have the type needed in the `typename IndexType`, no?

Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: that's not the type of the index, it's the type of the data

**Comment by mkrus** [comment]
added a simple test that data loads

File: src/core/gltf2importer/bufferaccessorparser.cpp (hunk 1 above)
Comment (jjcasmar) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: But arent you doing a switch to call this function and then doing another switch to call the other copySparseValues overload? My point is that at this point you dont need the switch because you can take directly sizeof(IndexType)

Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: you need 2 switches. The outer one is on the type of the index, the inner (in `copySparseValues`) is on the type of the values that get copied.

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 2]
```diff
@@ -106,6 +110,60 @@ Qt3DRender::QAttribute::VertexBaseType accessorTypeFromJSON(int componentType)
...
+    }
+    return true;
+}
+
+template<class IndexType>
+bool copySparseValues(int count, const QByteArray &indicesBuffer, const QByteArray &valuesBuffer,
+                      QByteArray &targetBuffer, int attributeOffset, int stride,
+                      Qt3DRender::QAttribute::VertexBaseType componentType, int numComponents)
+{
+    switch (componentType) {
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: wouldn't it be easier to read if in the switch we only computed the valued size allowing to then only have a single call to copySparseValues?

File: src/core/gltf2importer/animationparser.cpp [hunk 3]
```diff
@@ -88,19 +88,19 @@ int accessorDataTypeToBytes(Qt3DRender::QAttribute::VertexBaseType type)
 
 QByteArray rawDataFromAccessor(const Accessor &accessor, GLTF2Context *ctx)
 {
-    const BufferView &bufferViewData = ctx->bufferView(accessor.bufferViewIndex);
-
-    if (bufferViewData.bufferData.isEmpty()) {
+    if (accessor.bufferData.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (lemirep) [reply src/core/gltf2importer/animationparser.cpp]: Ah actually, accessor.bufferData is bufferViewData if we have a bufferViewIndex so that's fine.

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 4]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
         accessor.max = jsonArrayToVectorOfFloats(json.value(KEY_MAX).toArray());
         accessor.min = jsonArrayToVectorOfFloats(json.value(KEY_MIN).toArray());
         accessor.name = json.value(KEY_NAME).toString();
+
+        if (json.contains(KEY_SPARSE)) {
+            QJsonObject sparse = json.value(KEY_SPARSE).toObject();
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: const QJsonObject

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 5]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
...
+            if (!sparse.contains(KEY_COUNT) || !sparse.contains(KEY_SPARSE_INDICES) || !sparse.contains(KEY_SPARSE_VALUES)) {
+                qCWarning(kuesa, "Missing sparse data in accessor");
+                return false;
+            }
+
+            accessor.sparseCount = sparse.value(KEY_COUNT).toInt(0);
+
+            QJsonObject sparseIndices = sparse.value(KEY_SPARSE_INDICES).toObject();
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: const QJsonObject

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 6]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
...
+            accessor.sparseIndices.type = accessorTypeFromJSON(sparseIndices.value(KEY_COMPONENTTYPE).toInt(GL_UNSIGNED_INT));
+            accessor.sparseIndices.bufferViewIndex = sparseIndices.value(KEY_BUFFERVIEW).toInt(-1);
+            accessor.sparseIndices.offset = sparseIndices.value(KEY_BYTEOFFSET).toInt(0);
+
+            QJsonObject sparseValues = sparse.value(KEY_SPARSE_VALUES).toObject();
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: const QJsonObject

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 7]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
...
+            accessor.sparseValues.bufferViewIndex = sparseValues.value(KEY_BUFFERVIEW).toInt(-1);
+            accessor.sparseValues.offset = sparseValues.value(KEY_BYTEOFFSET).toInt(0);
+
+            if (accessor.sparseIndices.bufferViewIndex == -1 || accessor.sparseValues.bufferViewIndex == -1) {
+                qCWarning(kuesa, "Missing buffers for sparse indices or values in accessor");
+                return false;
+            }
+
+            QByteArray sparseIndicesData = context->bufferView(accessor.sparseIndices.bufferViewIndex).bufferData;
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: const QByteArray

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 8]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
...
+            if (sparseIndicesData.isEmpty() || sparseValuesData.isEmpty()) {
+                qCWarning(kuesa, "Empty buffers for sparse indices or values in accessor");
+                return false;
+            }
+
+            int stride = 0;
+            if (accessor.bufferData.isEmpty())
+                accessor.bufferData = QByteArray(accessor.dataSize * accessor.count, 0);
+            else {
+                accessor.bufferData.detach();
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: I'm not sure I understand why there's a detach here.

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 9]
```diff
@@ -135,6 +195,86 @@ bool Kuesa::GLTF2Import::BufferAccessorParser::parse(const QJsonArray &bufferAcc
...
+                const auto &bufferView = context->bufferView(accessor.bufferViewIndex);
+                stride = bufferView.byteStride;
+            }
+
+            switch (accessor.sparseIndices.type) {
+            case Qt3DRender::QAttribute::UnsignedShort:
```
Comment (lemirep) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: shouldn't the Byte/UnsignedByte types also be handled? specs seem to say it's possible unless I've misunderstood

File: tests/auto/gltfparser/tst_gltfparser.cpp [hunk 10]
```diff
@@ -855,6 +855,27 @@ private Q_SLOTS:
         // THEN
         QVERIFY(res != nullptr);
     }
+
+    void checkSparseAccessors()
+    {
```
Comment (lemirep) [inline tests/auto/gltfparser/tst_gltfparser.cpp]: Would it make sense to test against a file that has a sparse accessor that doesn't reference a bufferview to ensure the 0 filled generation works?

File: src/core/gltf2importer/bufferaccessorparser.cpp (hunk 2 above)
Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: good call

File: src/core/gltf2importer/bufferaccessorparser.cpp (hunk 8 above)
Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: we're going to modify the values in the bufferView, but the original content might be used in some other accessor. Sparse values are an optimisation to only store (in the gltf file) the delta between the original and modified content. At runtime you need both states

File: src/core/gltf2importer/bufferaccessorparser.cpp (hunk 9 above)
Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: oops

File: tests/auto/gltfparser/tst_gltfparser.cpp (hunk 10 above)
Comment (mkrus) [reply tests/auto/gltfparser/tst_gltfparser.cpp]: done

Comment (mkrus) [reply tests/auto/gltfparser/tst_gltfparser.cpp]: done

File: src/core/gltf2importer/bufferaccessorparser.cpp [hunk 11]
```diff
@@ -42,6 +42,7 @@ QT_BEGIN_NAMESPACE
 using namespace Kuesa;
 using namespace GLTF2Import;
 
+
```
Comment (jjcasmar) [inline src/core/gltf2importer/bufferaccessorparser.cpp]: extra line?

Comment (mkrus) [reply src/core/gltf2importer/bufferaccessorparser.cpp]: gone

File: tests/auto/gltfparser/tst_gltfparser.cpp [hunk 12]
```diff
@@ -855,6 +855,45 @@ private Q_SLOTS:
...
+
+        auto accessor = parser.context()->accessor(1);
+        QCOMPARE(accessor.sparseCount, 3);
+        QCOMPARE(accessor.sparseValues.bufferViewIndex, 3);
+        QCOMPARE(accessor.sparseIndices.bufferViewIndex, 2);
+        QVERIFY(!accessor.bufferData.isEmpty());
+
+        QCOMPARE(accessor.bufferViewIndex, bufferView);
+        const float *data = reinterpret_cast<const float*>(accessor.bufferData.constData());
+        QCOMPARE(data[3], originalValue);
```
Comment (lemirep) [inline tests/auto/gltfparser/tst_gltfparser.cpp]: I get a crash when exiting that test function for "Without Original BufferView" (not always seems to happen randomly). It seems there's a double free occurring sometimes when the SceneEntity gets destroyed. I'm not sure how this patch actually affects this.
